File: app/src/main/java/com/abs/launcher/model/FolderInfo.java
package com.abs.launcher.model;

import com.abs.launcher.ItemAction;
import com.abs.launcher.ItemInfo;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ZY on 2017/4/2.
 */

public class FolderInfo extends ItemInfo {

    ArrayList<ItemInfo> folderItems = new ArrayList<>();

    @Override
    public List<ItemAction> getDeleteAction(List<ItemAction> actions) {
        if (actions == null) {
            actions = new ArrayList<>();
        }
        return actions;
    }
}

File: app/src/main/java/com/abs/launcher/model/FolderInfo.kt
package com.abs.launcher.model

/**
 * Created by zy on 17-12-17.
 */
class FolderInfo : HomeItemInfo(ITEM_TYPE_FOLDER) {

    var title: String? = null
        get() {
            if (field == null) {
                //TODO get default folder name from resource
            }
            return field
        }

    private val contents: MutableList<FolderItemInfo> = ArrayList()

    override fun getActionList(): List<ItemAction> {
        return ArrayList()
    }

    fun add(info: FolderItemInfo) {
        info.getInfo().position.container = id
        for ((index, item) in contents.withIndex()) {
            if (item.getInfo().position.cellX >= info.getInfo().position.cellX) {
                contents.add(index, info)
                return
            }
        }
        contents.add(info)
    }

    fun getContents(): List<FolderItemInfo> {
        return ArrayList(contents)
    }

    fun isEmpty(): Boolean {
        return contents.isEmpty()
    }
}

interface FolderItemInfo {
    fun getInfo(): HomeItemInfo
}